### PR TITLE
Remove basic feature checks from license state

### PR DESCRIPTION
**Pull request by rjernst**
The XPackLicenseState is a utility to handle checking the currently
configured license level against the required license level of each
licensed feature. Originally all licensed features were paid features
and licenses always had a limited time scope. However, since the basic
license was introduced, all users have an unlimited time use of basic
features; there is no actual "basic license", rather it is the base case
for the default distribution. Therefore, the checks on basic features in
license state are no-ops because the current license always at least
allows basic features, and the basic license cannot be expired.

This commit removes all of the features from the license state marked as
BASIC or MISSING (a level lower than BASIC that also predates the basic
license and is no longer relevant). It also adds an assertion that no
basic license features can be added to the license state in the future.

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java [hunk 1]
```diff
@@ -69,17 +69,6 @@ public SecurityActionFilter(AuthenticationService authcService, AuthorizationSer
...
-        /*
-         A functional requirement - when the license of security is disabled (invalid/expires), security will continue
-         to operate normally, except all read operations will be blocked.
-         */
-        if (licenseState.checkFeature(Feature.SECURITY_STATS_AND_HEALTH) == false && LICENSE_EXPIRATION_ACTION_MATCHER.test(action)) {
-            logger.error("blocking [{}] operation due to expired license. Cluster health, cluster stats and indices stats \n" +
-                    "operations are blocked on license expiration. All data operations (read and write) continue to work. \n" +
-                    "If you have a new license, please update it. Otherwise, please reach out to your support contact.", action);
-            throw LicenseUtils.newComplianceException(XPackField.SECURITY);
-        }
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java:81]: My understanding on the intention of this piece of code is that we do not encourage users to run with expired commerical licenses, e.g. PLATINUM, forever. An expired PLATINUM license is still allowed to access some platinum features, e.g. DLS/FLS, for security reasons. If an expired license can access all BASIC featues, plus a fair bit more, this would put our licensing in an awkward situation. Therefore, we disable stats and health APIs to prompt users to either downgrade to a BASIC or renew the subscription.

In fact, many of the BASIC features have similar behaviour as the stats and health API, e.g. `ROLLUP(OperationMode.MISSING, true)`. The feature is available in BASIC, but also requires the license to be active.

**Review by ywangd — COMMENTED** [review]
I have one main concern:

Currently many of the basic features require the license to be active. My understanding is that we do not encourage users to run with expired commerical licenses forever. An expired PLATINUM license is still allowed to access some platinum features, e.g. DLS/FLS, for security reasons. If an expired license can access all BASIC featues, plus a fair bit more, this would put our licensing in an awkward situation. Therefore, we disable some of the BASIC features so that user can choose to either downgrade to a BASIC license or renew the subscription. With changes in this PR, an expired PLATINUM license would be all around better than a BASIC. Is this our intention?

I also have a few other less important suggestions:

* With this change, there are many `licenseState` fields that can be removed from various TransportAction classes, e.g. `RollupInfoTransportAction` as well as `ApiKeyService`.
* There seems to be opportunities for extracting common behaviours. For example, the `XxxInfoTransportAction` classes, since BASIC features can no longer be dsiabled and is always allowed, maybe we could have a superclass such as `BasicFeatureInfoTransportAction`. Security is a notable anomaly here. I wish it can also be always enabled, but that's a separate story.
* For the `Usage` classes, e.g. `RollupFeatureSetUsage`, I wonder whether we could add another constructor:
```java
public RollupFeatureSetUsage() {
    super(XPackField.ROLLUP, true, true);
}
```
and mark the current constructor `public RollupFeatureSetUsage(boolean available)` as deprecated or  just remove it. So we are not relying on the callers to pass in the right value.

**Comment by rjernst** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by rjernst** [comment]
@ywangd I've brought this PR back up to date and addressed your comments. Can you take another look?

Re your broader comment:
> With changes in this PR, an expired PLATINUM license would be all around better than a BASIC. Is this our intention?

Yes, this is the intention.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/rest/action/user/RestHasPrivilegesActionTests.java [hunk 2]
```diff
@@ -38,6 +40,7 @@
      */
     public void testBodyConsumed() throws Exception {
         final XPackLicenseState licenseState = mock(XPackLicenseState.class);
+        when(licenseState.isSecurityEnabled()).thenReturn(true);
```
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/rest/action/user/RestHasPrivilegesActionTests.java:43]: To retain the existing behaviour of this test, should this mock return `false` instead of `true` so that `action.handleRequest(...)` does not throw exception but returns a response about security is not enabled (currently it returns security not available due to license)?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/SqlUsageTransportAction.java [hunk 3]
```diff
@@ -58,7 +57,7 @@ protected void masterOperation(Task task, XPackUsageRequest request, ClusterStat
                 .filter(Objects::nonNull)
                 .collect(Collectors.toList());
             Counters mergedCounters = Counters.merge(countersPerNode);
-            SqlFeatureSetUsage usage = new SqlFeatureSetUsage(available, mergedCounters.toNestedMap());
+            SqlFeatureSetUsage usage = new SqlFeatureSetUsage(true, mergedCounters.toNestedMap());
```
Comment (ywangd) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/SqlUsageTransportAction.java:60]: Yet another `available` parameter.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/TransformUsageTransportAction.java [hunk 4]
```diff
@@ -96,7 +90,7 @@ protected void masterOperation(
         }
 
         ActionListener<TransformIndexerStats> totalStatsListener = ActionListener.wrap(statSummations -> {
-            var usage = new TransformFeatureSetUsage(available, transformsCountByState, statSummations);
+            var usage = new TransformFeatureSetUsage(true, transformsCountByState, statSummations);
```
Comment (ywangd) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/TransformUsageTransportAction.java:93]: Here again, the `available` parameter.

File: x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/VectorsUsageTransportAction.java [hunk 5]
```diff
@@ -76,7 +70,7 @@ protected void masterOperation(Task task, XPackUsageRequest request, ClusterStat
             }
         }
         VectorsFeatureSetUsage usage =
-            new VectorsFeatureSetUsage(vectorsAvailable, numDenseVectorFields, avgDenseVectorDims);
+            new VectorsFeatureSetUsage(true, numDenseVectorFields, avgDenseVectorDims);
```
Comment (ywangd) [inline x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/VectorsUsageTransportAction.java:73]: The `available` parameter again.

File: x-pack/plugin/voting-only-node/src/main/java/org/elasticsearch/cluster/coordination/VotingOnlyNodeFeatureSet.java [hunk 6]
```diff
@@ -53,36 +43,26 @@ public boolean enabled() {
...
-            this.licenseState = licenseState;
         }
 
         @Override
         protected void masterOperation(Task task, XPackUsageRequest request, ClusterState state,
                                        ActionListener<XPackUsageFeatureResponse> listener) {
-            final boolean available = licenseState.checkFeature(Feature.VOTING_ONLY);
-            final VotingOnlyNodeFeatureSetUsage usage =
-                new VotingOnlyNodeFeatureSetUsage(available);
+            final VotingOnlyNodeFeatureSetUsage usage = new VotingOnlyNodeFeatureSetUsage(true);
```
Comment (ywangd) [inline x-pack/plugin/voting-only-node/src/main/java/org/elasticsearch/cluster/coordination/VotingOnlyNodeFeatureSet.java:56]: I believe this is the last occurrence of the `available` parameter.

**Review by ywangd — APPROVED** [review]
LGTM. I left some comments, nothing really major.

> > With changes in this PR, an expired PLATINUM license would be all around better than a BASIC. Is this our intention?
>>
> Yes, this is the intention.
>
Well almost, the health and stats related actions are still blocked for expired licenses. So it still creates some prompt for users to choose either renew license or explicitly go back to basic.

I kinda like having a central list where every feature is explicitly listed along with its license requirement. It serves as a good reference. If we configure the license requirement correctly, something like `...(OperationMode.MISSING, false)`, we can even leave all the call sites of `checkFature` unchanged and just change the method itself to not track usages for any features of BASIC or lower. It also enforces license requirement be specified in one place so there is no place for inconsistency (among different classes usage/info/action etc). With that being said, I do like how much code we can delete because of the relevant Feature Enums are deleted. Given people who develop any new feature should also be responsible or consciously involved in license decision, I think it is reasonably to expect consistency will be taken care of. Overall I am fine with this approach.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/rest/action/user/RestHasPrivilegesActionTests.java (hunk 2 above)
Comment (ywangd) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/rest/action/user/RestHasPrivilegesActionTests.java:43]: We discussed this in another channel. After another look, I think the change is in fact closer to what this test tries to assert:
1. Send a request without an username
2. It should _not_ fail with IllegalArgumentException complaining about unconsumed payload

Both the existing version and the change do the job because both of them will run through `BaseRestHandler#handleRequest`, where the unconsumed body is checked.
The existing version does its job by ensure there is no exception. This relies on the fact that “security disabled” message takes precedence over the “no authenticated user” exception. The new change enables security so that the “no authenticated user” exception now bubbles up. Both “security disabled” message and “no authenticated user” exception have lower priority than the “unconsumed body” exception. So the test does its job if either of them can be asserted. But the new change has the advantage of an additional assertion on the exception to be “no authenticated user”, which is more to the point.

In summary, the new change does its job slightly differently than the existing version. But I think it's a small improvement. So we can discard my previous comment.
